pkg/cloud/buckets/buckets_test.go

```diff
@@ -27,3 +46,179 @@ func assertSplitBucketURL(t *testing.T, inputURL, expectedBucketURL, expectedKey
+					return &http.Response{
+						StatusCode: 200,
+						Body:       ioutil.NopCloser(bytes.NewReader([]byte(`[{"body": "ok"}]`))),
+					}, nil
+				},
+			},
+			ExpectedReadResponse: []byte(`[{"body": "ok"}]`),
+		},
+		{
+			ctx:           nil,
+			urlText:       "https://jenkins-x-demo.io/",
+			httpFn:        nil,
+			timeout:       30 * time.Second,
+			expectedError: fmt.Errorf("status Not found when performing GET on https://jenkins-x-demo.io/"),
+			httpResponse: &HTTPClientMock{
+				DoFunc: func(*http.Request) (*http.Response, error) {
+					return &http.Response{
+						Status:     "Not found",
+						StatusCode: 404,
+					}, nil
+				},
+			},
+		},
+		{
+			ctx:           nil,
+			urlText:       "https://jenkins-x-error.io/",
+			httpFn:        nil,
+			timeout:       30 * time.Second,
+			expectedError: fmt.Errorf("failed to invoke GET on https://jenkins-x-error.io/: No Internet Connection"),
+			httpResponse: &HTTPClientMock{
+				DoFunc: func(*http.Request) (*http.Response, error) {
+					return nil, fmt.Errorf("No Internet Connection")
+				},
+			},
+		},
+	}
+	for _, tc := range testCases {
+		if tc.httpResponse != nil {
+			buckets.HttpClient = tc.httpResponse
+			buckets.GetClientWithTimeout = func(timeout time.Duration) buckets.HTTPClient {
+				return buckets.HttpClient
+			}
+		}
+		read, err := buckets.ReadURL(tc.ctx, tc.urlText, tc.timeout, tc.httpFn)
+		if read != nil {
+			readResponse, err := ioutil.ReadAll(read)
+			assert.NoError(t, err)
+			assert.Equal(t, readResponse, tc.ExpectedReadResponse)
+		} else {
+			assert.Nil(t, read)
+		}
+		if err != nil {
+			assert.EqualError(t, err, tc.expectedError.Error())
+		} else {
+			assert.NoError(t, err)
+		}
+	}
+
+	ctx := context.Background()
+	read, err := buckets.ReadURL(ctx, "file:///test_data/foo", 30, nil)
+	assert.Nil(t, read)
+	assert.Contains(t, err.Error(), "failed to read key")
+}
+
+func TestBucketURL(t *testing.T) {
+	ctx := context.Background()
+	testCases := []struct {
+		url           string
+		data          string
+		time          time.Duration
+		expectedError string
+	}{
+		{
+			url:  "file:///test_data/my-key",
+			data: "Hello world",
+			time: 30 * time.Second,
+		},
+		{
+			url:           "x:///test_data/fooError",
+			data:          "Hello foo2",
+			time:          30 * time.Second,
+			expectedError: "failed to open bucket",
+		},
+	}
+	dir, err := os.Getwd()
+	assert.NoError(t, err)
+	defer os.RemoveAll(dir + "/test_data/")
+	for _, tc := range testCases {
+		u, err := url.Parse(tc.url)
+		assert.NoError(t, err)
+
+		err = buckets.WriteBucketURL(ctx, u, strings.NewReader(tc.data), tc.time)
+		if tc.expectedError != "" {
+			assert.Contains(t, err.Error(), "failed to open bucket")
+		} else {
+			assert.NoError(t, err)
+		}
+		read, err := buckets.ReadURL(ctx, tc.url, tc.time, nil)
+		if tc.expectedError != "" {
+			assert.Nil(t, read)
+			assert.Contains(t, err.Error(), "failed to open bucket")
```

Comment:
I think it's enough to test that there is an error, using contains is a bit brittle here.